core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -908,15 +921,37 @@ protected SearcherManager getSearcherManager() {
         return searcherManager;
     }
 
-    private Object dirtyLock(BytesRef uid) {
-        int hash = Murmur3HashFunction.hash(uid.bytes, uid.offset, uid.length);
-        return dirtyLocks[Math.floorMod(hash, dirtyLocks.length)];
+    private ReleasableLock dirtyLock(BytesRef uid) {
+        final Tuple<ReleasableLock, AbstractRefCounted> tuple = lockMap.compute(uid, (k, v) -> {
+            if (v == null || !v.v2().tryIncRef()) {
+                final ReleasableLock lock = locks.poll();
+                final AbstractRefCounted ref = new AbstractRefCounted(null) {
+                    @Override
+                    protected void closeInternal() {
+                        locks.offer(lock);
+                        lockMap.remove(k);
+                    }
+                };
+                return Tuple.tuple(lock, ref);
+            } else {
+                return v;
+            }
+        });
+        return tuple.v1();
     }
 
-    private Object dirtyLock(Term uid) {
+    private ReleasableLock dirtyLock(Term uid) {
         return dirtyLock(uid.bytes());
     }
 
+    private void remove(BytesRef uid) {
+        lockMap.get(uid).v2().decRef();
+    }
+
+    private void remove(Term uid) {
+        remove(uid.bytes());
+    }
+
```

Comment:
This logic looks very similar to KeyedLock, maybe we should use it instead?

I think the difference is that KeyedLock does not have a pool of locks and creates them on demand but I don't think reusing matters much here since we run costly operations under the lock?